engine/src/state_chain_observer/sc_observer/mod.rs

```diff
@@ -525,6 +516,39 @@ where
     }.instrument(info_span!("SCObserver")).boxed()).await
 }
 
+/// Get the CFE events from the SC at the given block with version checking
+pub async fn get_cfe_events(
+	state_chain_client: Arc<impl StorageApi>,
+	block_info: BlockInfo,
+) -> Result<Vec<CfeEvent>, anyhow::Error> {
+	// First we check the version number of the CFE events
+	let cfe_events_version = state_chain_client
+		.storage_value::<pallet_cf_cfe_interface::CfeEventsVersion<Runtime>>(block_info.hash)
+		.await
+		.expect("Failed to get CFE events version");
+
+	// For now we only have one version of CFE events, so we panic if the version number is not a
+	// match.
+	if cfe_events_version != CFE_EVENTS_VERSION {
+		panic!("Cannot process CFE events at block {}. Incompatible CFE events version: {cfe_events_version}. Using version: {CFE_EVENTS_VERSION}", block_info.number);
+	}
+
+	// Getting the events without decoding yet, so in the future we can decode them with the correct
+	// struct for the detected version.
+	match state_chain_client
+		.storage_value_raw::<pallet_cf_cfe_interface::CfeEvents<Runtime>>(block_info.hash)
```

Comment:
Think this is a bug in the same way we had the Polkadot bug. We should be getting the version at the *previous* block? 

I.e. events that are version 2, change to version 3 in block 20. The blocks *in* block 20 are version 2 but the blocks in block 21 are now version 3. And querying returns the value at the *end* of the block, if we queried for block 20, we'd get version 3, but the events within that block are version 2 events.

Would be nice if we could write a test for this.

Reply:
I'm not sure if this is true. Both the version and the events come from storage, and storage migration happen at the beginning of the block, I believe? Worth checking of course, but I would say a manual test with a runtime upgrade would suffice?